Bottom: bd94596808bc8239fa12c660fe278777d193e85e
Top:    5b5e9db3d94df4c32490b5dc29c887400922bf3f
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-18 19:51:51 +0100

Refresh of give-the-hids-control-over-att

---

diff --git a/src/hid/gtk/gtkhid-gdk.c b/src/hid/gtk/gtkhid-gdk.c
index e72aa65..3d90c27 100644
--- a/src/hid/gtk/gtkhid-gdk.c
+++ b/src/hid/gtk/gtkhid-gdk.c
@@ -34,6 +34,8 @@ typedef struct render_priv {
   GdkGC *mask_gc;
   GdkGC *u_gc;
   GdkGC *grid_gc;
+  int attached_invalidate_depth;
+  int mark_invalidate_depth;
 } render_priv;
 
 
@@ -709,6 +711,7 @@ ghid_invalidate_lr (int left, int right, int top, int bottom)
   ghid_invalidate_all ();
 }
 
+
 void
 ghid_invalidate_all ()
 {
@@ -771,64 +774,67 @@ ghid_invalidate_all ()
   hid_expose_callback (&ghid_hid, &region, 0);
   ghid_draw_grid ();
 
-  DrawAttached ();
-  DrawMark ();
+  if (priv->attachment_invalidate_depth == 0)
+    DrawAttached ();
+  if (priv->mark_invalidate_depth == 0)
+    DrawMark ();
 
   ghid_screen_update ();
 }
 
+
 void
 ghid_notify_crosshair_change (bool changes_complete)
 {
-  static int invalidate_depth = 0;
+  render_priv *priv = gport->render_priv;
 
-  /* FIXME: We sometimes get called before the GUI is up */
-  if (gport->drawing_area == NULL)
+  /* We sometimes get called before the GUI is up */
+  if (priv->drawing_area == NULL)
     return;
 
   if (changes_complete)
-    invalidate_depth --;
+    priv->attached_invalidate_depth --;
 
-  if (invalidate_depth < 0)
+  if (priv->attached_invalidate_depth < 0)
     {
       fprintf (stderr, "ERROR: Unmatched notify_crosshair_change calls\n");
-      invalidate_depth = 0;
+      priv->attached_invalidate_depth = 0;
     }
 
-  if (invalidate_depth == 0)
+  if (priv->attached_invalidate_depth == 0)
     DrawAttached ();
 
   if (!changes_complete)
-    invalidate_depth ++;
-  else if (gport->drawing_area != NULL)
-    ghid_draw_area_update (gport, NULL);
+    priv->attached_invalidate_depth ++;
+  else if (priv->drawing_area != NULL)
+    ghid_draw_area_update (priv, NULL);
 }
 
 void
 ghid_notify_mark_change (bool changes_complete)
 {
-  static int invalidate_depth = 0;
+  render_priv *priv = gport->render_priv;
 
-  /* FIXME: We sometimes get called before the GUI is up */
-  if (gport->drawing_area == NULL)
+  /* We sometimes get called before the GUI is up */
+  if (priv->drawing_area == NULL)
     return;
 
   if (changes_complete)
-    invalidate_depth --;
+    priv->mark_invalidate_depth --;
 
-  if (invalidate_depth < 0)
+  if (priv->mark_invalidate_depth < 0)
     {
       fprintf (stderr, "ERROR: Unmatched notify_mark_change calls\n");
-      invalidate_depth = 0;
+      priv->mark_invalidate_depth = 0;
     }
 
-  if (invalidate_depth == 0)
+  if (priv->mark_invalidate_depth == 0)
     DrawMark ();
 
   if (!changes_complete)
-    invalidate_depth ++;
+    priv->mark_invalidate_depth ++;
   else
-    ghid_draw_area_update (gport, NULL);
+    ghid_draw_area_update (priv, NULL);
 }
 
 static void
diff --git a/src/hid/gtk/gui-output-events.c b/src/hid/gtk/gui-output-events.c
index 791a608..03fa2da 100644
--- a/src/hid/gtk/gui-output-events.c
+++ b/src/hid/gtk/gui-output-events.c
@@ -58,6 +58,9 @@ ghid_port_ranges_changed (void)
 {
   GtkAdjustment *h_adj, *v_adj;
 
+  HideCrosshair ();
+  ghidgui->need_restore_crosshair = TRUE;
+
   h_adj = gtk_range_get_adjustment (GTK_RANGE (ghidgui->h_range));
   v_adj = gtk_range_get_adjustment (GTK_RANGE (ghidgui->v_range));
   gport->view_x0 = h_adj->value;
@@ -491,6 +494,7 @@ ghid_port_drawing_area_configure_event_cb (GtkWidget * widget,
 {
   static gboolean first_time_done;
 
+  HideCrosshair ();
   gport->width = ev->width;
   gport->height = ev->height;
 
@@ -524,6 +528,7 @@ ghid_port_drawing_area_configure_event_cb (GtkWidget * widget,
 
   ghid_port_ranges_scale (FALSE);
   ghid_invalidate_all ();
+  RestoreCrosshair ();
   return 0;
 }
